### PR TITLE
Support Amazon Linux 2023

### PR DESCRIPTION
**Pull request by himani2411**
### Description of changes
* See commit descriptions for details

### Tests
AMIs build are successful. All integration tests in develop.yaml have been run.
Only the following 6 tests are **failing**:
```
test-suites:
  basic:
    test_essential_features.py::test_essential_features:
      dimensions:
      - instances:
        - c5.xlarge
        oss:
        - alinux2023
        regions:
        - us-east-1
        schedulers:
        - slurm
  createami:
    test_createami.py::test_build_image_custom_components:
      dimensions:
      - instances:
        - m6g.xlarge
        oss:
        - alinux2023
        regions:
        - us-west-1
      - instances:
        - c5.xlarge
        oss:
        - alinux2023
        regions:
        - eu-north-1
  iam:
    test_iam_image.py::test_iam_roles:
      dimensions:
      - instances:
        - c5.xlarge
        oss:
        - alinux2023
        regions:
        - us-west-2
  networking:
    test_cluster_networking.py::test_cluster_in_no_internet_subnet:
      dimensions:
      - instances:
        - c5.xlarge
        oss:
        - alinux2023
        regions:
        - us-east-1
        schedulers:
        - slurm
  trainium:
    test_trainium.py::test_trainium:
      dimensions:
      - oss:
        - alinux2023
        regions:
        - us-west-2
        schedulers:
        - slurm

```

### References
* Cookbook PR: https://github.com/aws/aws-parallelcluster-cookbook/pull/2692

### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,7 @@ CHANGELOG
 
 **ENHANCEMENTS**
 - Add new configuration section `Scheduling/SlurmSettings/ExternalSlurmdbd` to connect the cluster to an external Slurmdbd.
+Add support for Amazon Linux 2023.
```
Comment (gmarciani) [inline CHANGELOG.md]: typo: missing '-'

File: cli/src/pcluster/templates/cluster_stack.py [hunk 2]
```diff
@@ -1398,8 +1398,7 @@ def _add_head_node(self):
                     "touch": {"command": "touch /etc/chef/ohai/hints/ec2.json"},
                     "jq": {
                         "command": (
-                            "jq --argfile f1 /tmp/dna.json --argfile f2 /tmp/extra.json -n '$f1 * $f2' "
-                            "> /etc/chef/dna.json "
+                            'jq -s ".[0] * .[1]" /tmp/dna.json /tmp/extra.json > /etc/chef/dna.json '
```
Comment (gmarciani) [inline cli/src/pcluster/templates/cluster_stack.py:1401]: How is this change related to the support for AL2023?

File: tests/integration-tests/configs/new_os.yaml [hunk 3]
```diff
@@ -72,6 +77,7 @@ test-suites:
     test_createami.py::test_build_image:
       dimensions:
         - regions: ["eu-west-3"]
+          schedulers: [ "slurm" ]
```
Comment (gmarciani) [inline tests/integration-tests/configs/new_os.yaml:80]: Why is this change required now?

File: tests/integration-tests/tests/arm_pl/test_arm_pl.py [hunk 4]
```diff
@@ -30,12 +30,12 @@ def test_arm_pl(os, pcluster_config_reader, clusters_factory, test_datadir):
...
     # arm performance library version and gcc version
-    armpl_version = "23.04.1" if os == "ubuntu2204" else "21.0.0"
-    gcc_version = "11.3" if os == "ubuntu2204" else "9.3"
+    armpl_version = "23.10.0"
+    gcc_version = "11.3" if os in ["ubuntu2204", "alinux2023", "redhat9", "rocky9"] else "9.3"
 
     # loading module armpl/{armpl_version} will load module armpl/gcc-{gcc_version}
     # and armpl/{armpl_version}_gcc-{gcc_vesion}  sequentially
-    armpl_module_general_name = f"armpl/{armpl_version}"
+    armpl_module_general_name = f"armpl/{armpl_version[:-2]}" if os in ["alinux2023"] else f"armpl/{armpl_version}"
```
Comment (gmarciani) [inline tests/integration-tests/tests/arm_pl/test_arm_pl.py:38]: Minor: in the comment above add an example of how you expect the string to be for AL2023.

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 5]
```diff
@@ -1793,9 +1793,7 @@ def _test_torque_job_submit(remote_command_executor, test_datadir):
 def _submit_kill_networking_job(remote_command_executor, scheduler_commands, partition, node_type, num_nodes):
     """Submit job that will detach network interface on compute."""
     # Get network interface name from Head node, assuming Head node and Compute are of the same instance type
-    interface_name = remote_command_executor.run_remote_command(
-        "nmcli device status | grep ether | awk '{print $1}'"
-    ).stdout
+    interface_name = remote_command_executor.run_remote_command("ls /sys/class/net | grep e").stdout
```
Comment (gmarciani) [inline tests/integration-tests/tests/schedulers/test_slurm.py:1796]: why is this changing?

File: cli/src/pcluster/templates/cluster_stack.py (hunk 2 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/templates/cluster_stack.py:1401]: New versions of jq deprecated `--argfile`

File: tests/integration-tests/configs/new_os.yaml (hunk 3 above)
Comment (hanwen-pcluste) [reply tests/integration-tests/configs/new_os.yaml:80]: This change is optional

**Review by hanwen-pcluste — DISMISSED** [review]

File: tests/integration-tests/tests/schedulers/test_slurm.py (hunk 5 above)
Comment (hanwen-pcluste) [reply tests/integration-tests/tests/schedulers/test_slurm.py:1796]: nmcli is deprecated by amazon linux 2023

File: CHANGELOG.md (hunk 1 above)
Comment (hanwen-pcluste) [reply CHANGELOG.md]: Done

File: tests/integration-tests/configs/new_os.yaml (hunk 3 above)
Comment (himani2411) [reply tests/integration-tests/configs/new_os.yaml:80]: Minor: This change is something that we missed when adding this new_os.yaml, if this is not added the test fails with missing schedulers dimensions

File: cli/src/pcluster/resources/batch/docker/alinux2023/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,60 @@
+FROM public.ecr.aws/amazonlinux/amazonlinux:2023
+
+ENV USER root
+
+ARG AWS_REGION
```
Comment (enrico-usai) [inline cli/src/pcluster/resources/batch/docker/alinux2023/Dockerfile:5]: We're adding alinux2023 as base OS for AWS Batch, did we test it using alinux2023 for batch tests?
If so we should mention this in the changelog and bump the version of awsbatch cli.
Otherwise we can avoid to create this dockerfile.

Comment (gmarciani) [reply cli/src/pcluster/resources/batch/docker/alinux2023/Dockerfile:5]: Good point! Thanks Enrico!

**Comment by gmarciani** [comment]
PR LGTM. Will approved once addressed this comment https://github.com/aws/aws-parallelcluster/pull/6223#discussion_r1634393396

File: cli/src/pcluster/resources/batch/docker/alinux2023/Dockerfile (hunk 6 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/resources/batch/docker/alinux2023/Dockerfile:5]: We tested with Batch and the following integration tests have passed. We will bump the version of Batch CLI during the release process
```
{%- import 'common.jinja2' as common with context -%}
test-suites:
  schedulers:
    test_awsbatch.py::test_awsbatch:
      dimensions:
        - regions: ["us-west-1"]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: ["alinux2023"]
          schedulers: ["awsbatch"]
    test_awsbatch.py::test_awsbatch_defaults:
      dimensions:
        - regions: ["ap-northeast-1"]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: ["alinux2023"]
          schedulers: ["awsbatch"]
```

**Review by gmarciani — DISMISSED** [review]

**Review by gmarciani — APPROVED** [review]
